### PR TITLE
feat: Added favorite item hotkey [skip ci]

### PR DESCRIPTION
**Pull request by aprevrha**
A hotkey that makes favoriting items faster.

By default the hotkey only works on gear items and ingredients but there is a configuration option that allows favoriting of any item via the hotkey.
image

### CODE REVIEW

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        gearItemOpt = Models.Item.asWynnItem(hoveredSlot.getItem(), GearItem.class);
+        if (gearItemOpt.isPresent()) itemName = gearItemOpt.get().getName();
+
+        ingredientItemOpt = Models.Item.asWynnItem(hoveredSlot.getItem(), IngredientItem.class);
+        if (ingredientItemOpt.isPresent()) itemName = ingredientItemOpt.get().getName();
+
+        if (allowFavoritingAllItems.get()) itemName = unformattedName;
+
+        if (itemName != null && !itemName.isBlank()) Services.Favorites.toggleFavorite(itemName);
```
Comment (ryanzhoudev) [inline common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java]: Fix styling (use { }) on all of these please - we only use one liners for simple flow control, eg. return/continue/break.

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java:1]: Can this not be added to the existing `ItemFavoriteFeature` why does it need to be separate?

Comment (magicus) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java:1]: I agree, that would be a much better fit.

**Comment by ShadowCat117** [comment]
If I remember correctly we have a `NamedItemProperty` which I believe all of the current guide items are instances of so this could be simplified a lot by checking if the current item is one of those. Then if the allow all favourites setting is enabled then you can just use the unformatted name

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java:1]: Yup agree, I'll do this

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    @Persisted
+    public final Config<Boolean> allowFavoritingAllItems = new Config<>(false);
+
+    private void tryChangeFavoriteStateOnHoveredSlot(Slot hoveredSlot) {
+        if (hoveredSlot == null || !hoveredSlot.hasItem()) return;
+
+        String unformattedName = PATTERN.matcher(
+                        StyledText.fromComponent((hoveredSlot.getItem().getHoverName()))
+                                .getStringWithoutFormatting())
+                .replaceAll("");
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java]: Will `WynnUtils#normalizeBadString` accomplish the same goal as this?

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java:1]: done.

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java (hunk 1 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java]: This should be fixed now, let me know if there is anything else

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java (hunk 2 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteHotkeyFeature.java]: Yeah it works and looks a lot nicer.

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java [hunk 3]
```diff
@@ -4,21 +4,27 @@
  */
 package com.wynntils.features.inventory;
 
+import static com.wynntils.utils.wynn.WynnUtils.normalizeBadString;
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: Don't use static imports

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java [hunk 4]
```diff
@@ -121,4 +137,22 @@ private static void renderFavoriteItem(SlotRenderEvent.Post event) {
...
+                Models.Item.asWynnItemProperty(hoveredSlot.getItem(), NamedItemProperty.class);
+        if (namedItemPropertyOpt.isPresent()) {
+            itemName = namedItemPropertyOpt.get().getName();
+        } else if (allowFavoritingAllItems.get()) {
+            itemName = normalizeBadString(
+                    StyledText.fromComponent((hoveredSlot.getItem().getHoverName()))
+                            .getStringWithoutFormatting());
+        }
+
+        if (itemName != null && !itemName.isBlank()) Services.Favorites.toggleFavorite(itemName);
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: Remember the brackets :)

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java (hunk 3 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: uh ok

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java (hunk 4 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: those sneaky one liners ay :)

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java (hunk 3 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: fixed :)

File: common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java (hunk 4 above)
Comment (antonprev) [reply common/src/main/java/com/wynntils/features/inventory/ItemFavoriteFeature.java]: Hope there are no more formatting errors. I'm new to this modding :D

**Review by ShadowCat117 — APPROVED** [review]
Looks good to me. Thank you for the contribution! :)

**Review by magicus — APPROVED** [review]
This looks great now! Thanks for the contribition, and for working with us to get the patch to this shape.

**Pull request merged**